tests/class-two-factor-core.php

```diff
@@ -862,6 +862,44 @@ public function test_enabling_two_factor_is_factored_session() {
 
 	}
 
+	/**
+	 * Validate that disabling all providers [invalidate] the two-factor session.
+	 *
+	 * @covers Two_Factor_Core::current_user_can_update_two_factor_options()
+	 * @covers Two_Factor_Core::user_two_factor_options_update()
+	 */
+	public function test_disabling_two_factor_is_not_factored_session() {
+		$user = self::factory()->user->create_and_get();
+
+		// Setup a 2FA session.
+		wp_set_current_user( $user->ID );
+		wp_set_auth_cookie( $user->ID );
+
+		$key              = '_nonce_user_two_factor_options';
+		$nonce            = wp_create_nonce( 'user_two_factor_options' );
+		$_POST[ $key ]    = $nonce;
+		$_REQUEST[ $key ] = $nonce;
+
+		$_POST[ Two_Factor_Core::ENABLED_PROVIDERS_USER_META_KEY ] = [ 'Two_Factor_Dummy' => 'Two_Factor_Dummy' ];
+
+		Two_Factor_Core::user_two_factor_options_update( $user->ID );
+
+		$this->assertNotFalse( Two_Factor_Core::is_current_user_session_two_factor() );
+
+		// Disable all providers, and test that the 2FA session is invalidated.
+		$_POST[ Two_Factor_Core::ENABLED_PROVIDERS_USER_META_KEY ] = [];
+		Two_Factor_Core::user_two_factor_options_update( $user->ID );
+
+		$this->assertFalse( Two_Factor_Core::is_current_user_session_two_factor() );
+
+		$manager = WP_Session_Tokens::get_instance( $user->ID );
+		$token   = wp_get_session_token();
+		$session = $manager->get( $token );
+
+		$this->assertArrayNotHasKey( 'two-factor-login', $session );
+		$this->assertArrayNotHasKey( 'two-factor-provider', $session );
```

Comment:
I guess at the time I thought something like "The fact the user is no longer logged in with 2FA doesn't suddenly make their session not a 2fa session", but somewhere along the line with the logic changes I've missed that they couldn't then update the session after disabling it..